File: 1846. maximum-element-after-decreasing-and-rearranging.py
from typing import List

class Solution:
    def maximumElementAfterDecrementingAndRearranging(self, arr: List[int]) -> int:

        arr.sort()

        arr[0] = 1

        for i in range(1, len(arr)):
            arr[i] = min(arr[i], arr[i - 1] + 1)

        return arr[-1]


if __name__ == "__main__":
    sol = Solution()

    print(sol.maximumElementAfterDecrementingAndRearranging([2, 2, 1, 2, 1]))  # Expected: 2
    print(sol.maximumElementAfterDecrementingAndRearranging([100, 1, 1000]))    # Expected: 3
    print(sol.maximumElementAfterDecrementingAndRearranging([1, 2, 3, 4, 5]))   # Expected: 5

File: 1846. maximum-element-after-decreasing-and-rearranging.java
import java.util.Arrays;

class Solution {
    public int maximumElementAfterDecrementingAndRearranging(int[] arr) {

        Arrays.sort(arr);

        arr[0] = 1;

        for (int i = 1; i < arr.length; i++) {

            arr[i] = Math.min(arr[i], arr[i - 1] + 1);
        }

        return arr[arr.length - 1];
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        System.out.println(sol.maximumElementAfterDecrementingAndRearranging(
            new int[]{2, 2, 1, 2, 1}
        )); // Expected: 2

        System.out.println(sol.maximumElementAfterDecrementingAndRearranging(
            new int[]{100, 1, 1000}
        )); // Expected: 3

        System.out.println(sol.maximumElementAfterDecrementingAndRearranging(
            new int[]{1, 2, 3, 4, 5}
        )); // Expected: 5
    }
}